### DISCUSSION TITLE
Suggested way to handle pytest fixtures not being accessed warning?

### DISCUSSION THREAD

**Discussion by Ran4** [microsoft/pyright#1738] [Q&A]
Most of my projects use pytest, and most use pytest fixtures. Essentially, fixtures is a function that pytest will call before calling the test. But since the calling happens inside of pytest and not as part of the test code itself, pyright warns about my fixture functions not being accessed.

E.g.

```python
# fixtures.py
import pytest

@pytest.fixture
def age() -> int:
    return 42
```

```python
# test_something.py
from fixtures import age

def something(age: int):
    return True

def test_something(age):
    assert something(age)
```

pyright will complain about the `from fixtures import age` line.

I do know that the pyright project doesn't want to implement third party workarounds, but is there any workarounds to avoid my editor complaining about these unused imports?

**Comment by erictraut** [comment]
By default, pyright doesn't emit any errors or warnings for unused import symbols, so I presume that you've enabled the `reportUnusedImport` diagnostic rule explicitly or enabled strict mode, which implicitly enables `reportUnusedImport`. 

So you have several options:
1. Don't enable the `reportUnusedImport` diagnostic check in your settings or config file.
2. Disable the check in specific files using a `# pyright: reportUnusedImport=false` comment at the top of the file.
3. Use a `# type: ignore` comment to suppress the diagnostic on that specific line.
4. Use the "redundant" form of the import statement: `from fixtures import age as age`, which treats the imported symbol as accessed, since it's considered a re-export.
5. Add a dummy usage of the imported symbol somewhere in the file (e.g. `_ = age`)

Alternatively (and this is the way my team handles this in our pytest-based tests), you can define your fixtures in a conftest.py file and configure pytest to consume them from this file. That way, you don't need to import them into each test module explicitly. For more details, refer to the [pytest documentation](https://docs.pytest.org/en/stable/fixture.html#scope-sharing-fixtures-across-classes-modules-packages-or-session).

**Reply by Ran4** [reply]
Thank you, this solved my problem!

These are my results, using coc-pyright

1. -
2. Doesn't work - I still get a warning. Probably an issue with coc-pyright, not pyright
3. Also doesn't work - I still get a warning. Probably an issue with coc-pyright, not pyright
4. Works!
5. Works! I think I'll be using this one, with a comment on why it's there (so other teammembers can understand it)
6. Works! I won't be going this route since I have multiple fixture functions with the same name, and getting the module resolution correct when having multiple conftest.py seems like a pain, but for most projects this is probably a good solution.

**Reply by erictraut** [reply]
If ideas 1 through 3 do not work, then you do not have the `reportUnusedImport` diagnostic enabled, and pyright is not issuing an error or warning. It is instead generating a "hint", which the editor is supposed to display not as an error or warning but rather as "grayed-out" text. If your editor is interpreting in another manner (e.g. underlining the text or displaying an error or warning), then it's a bug in your editor or the LSP, and you should report it as a bug. This works fine in VS Code.

So I'll add another solution for you to consider:
7. Switch to VS Code. :)

**Reply by jbethune** [reply]
> Alternatively (and this is the way my team handles this in our pytest-based tests), you can define your fixtures in a conftest.py file and configure pytest to consume them from this file. That way, you don't need to import them into each test module explicitly. For more details, refer to the [pytest documentation](https://docs.pytest.org/en/stable/fixture.html#scope-sharing-fixtures-across-classes-modules-packages-or-session).

Just an update for 2025: It seems that the documentation for `conftest.py` files has been moved here: https://docs.pytest.org/en/stable/how-to/fixtures.html#scope-sharing-fixtures-across-classes-modules-packages-or-session
